core/Pos/Core/Genesis/Types.hs

```diff
@@ -155,37 +204,41 @@ bootDustThreshold (GenesisWStakeholders bootHolders) =
     -- even more than 10-15 coins.
     unsafeIntegerToCoin . sum $ map fromIntegral $ toList bootHolders
 
--- | Safe constructor for 'GenesisCoreData'. Throws error if something
+-- | Safe constructor for 'GenesisSpec'. Throws error if something
 -- goes wrong.
-mkGenesisCoreData ::
-       [AddrDistribution]
-    -> Map StakeholderId Word16
+mkGenesisSpec
+    :: GenesisAddrDistr
     -> GenesisDelegation
-    -> Either String GenesisCoreData
-mkGenesisCoreData distribution bootStakeholders delega = do
-    for_ distribution $ \(addrs, distr) -> do
-        -- Every set of addresses should match the stakeholders count
-        unless (fromIntegral (length addrs) == getDistributionSize distr) $
-            Left "mkGenesisCoreData: addressCount != stakeholdersCount \
-                 \for some set of addresses"
-        -- Addresses in each list are distinct (except for CustomBalances)
-        let isCustom = case distr of
-                CustomBalances{} -> True
-                _                -> False
-        unless (isCustom || allDistinct addrs) $
-            Left "mkGenesisCoreData: addresses in some list aren't distinct"
-        unless (all isBootstrapEraDistrAddress addrs) $
-            Left $ "mkGenesisCoreData: there is an address with stake " <>
-                   "distribution different from BootstrapEraDistr"
-
-    -- No address belongs to more than one distribution
-    let addrList = concatMap (ordNub . fst) distribution
-    unless (allDistinct addrList) $
-        Left "mkGenesisCoreData: some address belongs to more than one distr"
+    -> GenesisInitializer
+    -> Either String GenesisSpec
+mkGenesisSpec avvmDistr delega specType = do
+    let addrs = HM.keys $ getAddrDistr avvmDistr
+    unless (all isBootstrapEraDistrAddress addrs) $
+        Left $ "mkGenesisCoreData: there is an address with stake " <>
+                "distribution different from BootstrapEraDistr"
+
     -- All checks passed
-    pure
-        UnsafeGenesisCoreData
-        { gcdAddrDistribution = distribution
-        , gcdBootstrapStakeholders = GenesisWStakeholders bootStakeholders
-        , gcdHeavyDelegation = delega
-        }
+    pure $ UnsafeGenesisSpec avvmDistr delega specType
+
+data ProtocolConstants = ProtocolConstants
+    {
+      -- | Security parameter from paper
+      pcK             :: !Int
+      -- | Magic constant for separating real/testnet
+    , pcProtocolMagic :: !Int32
```

Comment:
We'll run into trouble with this if we insist on treating it like a top-level constant via `newInitFuture`. Protocol magic is defined here in a module which depends upon `Pos.Crypto.SignTag`, but then `Pos.Crypto.SignTag` makes use of this faux top-level constant and so we get a cycle.

The natural way of fixing this is to have `signTag` take the protocol magic as a parameter.

Reply:
Breaking it with an `hs-boot` for `Pos.Core.Constants.Raw` should be ok I guess. The `hs-boot` will define `protocolMagic` only.